File: src/compiler/parse_grammar.cc
#include "compiler/parse_grammar.h"
#include <string>
#include <vector>
#include <unordered_set>
#include <utility>
#include "json.h"
#include "compiler/rule.h"

namespace tree_sitter {

using std::move;
using std::string;
using std::vector;
using std::unordered_set;
using std::pair;
using rules::Rule;
using rules::Blank;
using rules::Metadata;
using rules::Pattern;
using rules::String;
using rules::NamedSymbol;

struct ParseRuleResult {
  Rule rule;
  string error_message;

  ParseRuleResult(const string &error_message) : error_message(error_message) {}
  ParseRuleResult(const char *error_message) : error_message(error_message) {}
  ParseRuleResult(Rule rule) : rule(rule) {}
};

ParseRuleResult parse_rule(json_value *rule_json) {
  string error_message;
  json_value rule_type_json;
  string type;

  if (!rule_json) {
    return "Rule cannot be null";
  }

  if (rule_json->type != json_object) {
    return "Rule type must be an object";
  }

  rule_type_json = rule_json->operator[]("type");
  if (rule_type_json.type != json_string) {
    return "Rule type must be a string";
  }

  type = rule_type_json.u.string.ptr;

  if (type == "BLANK") {
    return Rule(Blank{});
  }

  if (type == "CHOICE") {
    json_value members_json = rule_json->operator[]("members");
    if (members_json.type != json_array) {
      return "Choice members must be an array";
    }

    vector<Rule> members;
    for (size_t i = 0, length = members_json.u.array.length; i < length; i++) {
      json_value *member_json = members_json.u.array.values[i];
      auto result = parse_rule(member_json);
      if (!result.error_message.empty()) {
        return "Invalid choice member: " + result.error_message;
      }
      members.push_back(result.rule);
    }
    return Rule::choice(members);
  }

  if (type == "SEQ") {
    json_value members_json = rule_json->operator[]("members");
    if (members_json.type != json_array) {
      return "Seq members must be an array";
    }

    vector<Rule> members;
    for (size_t i = 0, length = members_json.u.array.length; i < length; i++) {
      json_value *member_json = members_json.u.array.values[i];
      auto result = parse_rule(member_json);
      if (!result.error_message.empty()) {
        return "Invalid choice member: " + result.error_message;
      }
      members.push_back(result.rule);
    }
    return Rule::seq(members);
  }

  if (type == "REPEAT") {
    json_value content_json = rule_json->operator[]("content");
    auto result = parse_rule(&content_json);
    if (!result.error_message.empty()) {
      return "Invalid repeat content: " + result.error_message;
    }
    return Rule::choice({Rule::repeat(result.rule), Blank{}});
  }

  if (type == "REPEAT1") {
    json_value content_json = rule_json->operator[]("content");
    auto result = parse_rule(&content_json);
    if (!result.error_message.empty()) {
      return "Invalid repeat content: " + result.error_message;
    }
    return Rule::repeat(result.rule);
  }

  if (type == "TOKEN") {
    json_value content_json = rule_json->operator[]("content");
    auto result = parse_rule(&content_json);
    if (!result.error_message.empty()) {
      return "Invalid token content: " + result.error_message;
    }
      return Rule(Metadata::token(move(result.rule)));
  }

  if (type == "IMMEDIATE_TOKEN") {
    json_value content_json = rule_json->operator[]("content");
    auto result = parse_rule(&content_json);
    if (!result.error_message.empty()) {
      return "Invalid token content: " + result.error_message;
    }
      return Rule(Metadata::immediate_token(move(result.rule)));
  }

  if (type == "PATTERN") {
    json_value value_json = rule_json->operator[]("value");
    if (value_json.type == json_string) {
      return Rule(Pattern{value_json.u.string.ptr});
    } else {
      return "Pattern value must be a string";
    }
  }

  if (type == "STRING") {
    json_value value_json = rule_json->operator[]("value");
    if (value_json.type == json_string) {
      return Rule(String{value_json.u.string.ptr});
    } else {
      return "String rule value must be a string";
    }
  }

  if (type == "SYMBOL") {
    json_value value_json = rule_json->operator[]("name");
    if (value_json.type == json_string) {
      return Rule(NamedSymbol{value_json.u.string.ptr});
    } else {
      return "Symbol value must be a string";
    }
  }

  if (type == "PREC") {
    json_value precedence_json = rule_json->operator[]("value");
    if (precedence_json.type != json_integer) {
      return "Precedence value must be an integer";
    }

    json_value content_json = rule_json->operator[]("content");
    auto result = parse_rule(&content_json);
    if (!result.error_message.empty()) {
      return "Invalid precedence content: " + result.error_message;
    }
    return Rule(Metadata::prec(precedence_json.u.integer, move(result.rule)));
  }

  if (type == "PREC_LEFT") {
    json_value precedence_json = rule_json->operator[]("value");
    if (precedence_json.type != json_integer) {
      return "Precedence value must be an integer";
    }

    json_value content_json = rule_json->operator[]("content");
    auto result = parse_rule(&content_json);
    if (!result.error_message.empty()) {
      return "Invalid precedence content: " + result.error_message;
    }
    return Rule(Metadata::prec_left(precedence_json.u.integer, move(result.rule)));
  }

  if (type == "PREC_RIGHT") {
    json_value precedence_json = rule_json->operator[]("value");
    if (precedence_json.type != json_integer) {
      return "Precedence value must be an integer";
    }

    json_value content_json = rule_json->operator[]("content");
    auto result = parse_rule(&content_json);
    if (!result.error_message.empty()) {
      return "Invalid precedence content: " + result.error_message;
    }
    return Rule(Metadata::prec_right(precedence_json.u.integer, move(result.rule)));
  }

  if (type == "PREC_DYNAMIC") {
    json_value precedence_json = rule_json->operator[]("value");
    if (precedence_json.type != json_integer) {
      return "Precedence value must be an integer";
    }

    json_value content_json = rule_json->operator[]("content");
    auto result = parse_rule(&content_json);
    if (!result.error_message.empty()) {
      return "Invalid precedence content: " + result.error_message;
    }
    return Rule(Metadata::prec_dynamic(precedence_json.u.integer, move(result.rule)));
  }

  if (type == "ALIAS") {
    json_value value_json = rule_json->operator[]("value");
    if (value_json.type != json_string) {
      return "Rename value must be a string";
    }

    json_value is_named_json = rule_json->operator[]("named");
    if (is_named_json.type != json_boolean) {
      return "Rename named value must be a boolean";
    }

    json_value content_json = rule_json->operator[]("content");
    auto result = parse_rule(&content_json);
    if (!result.error_message.empty()) {
      return "Invalid rename content: " + result.error_message;
    }
    return Rule(Metadata::alias(
      string(value_json.u.string.ptr),
      is_named_json.u.boolean,
      move(result.rule)
    ));
  }

  return "Unknown rule type: " + type;
}

ParseGrammarResult parse_grammar(const string &input) {
  string error_message;
  string name;
  InputGrammar grammar;
  json_value
    name_json, rules_json, extras_json, conflicts_json, external_tokens_json,
    inline_rules_json, word_rule_json;

  json_settings settings = { 0, json_enable_comments, 0, 0, 0, 0 };
  char parse_error[json_error_max];
  json_value *grammar_json =
    json_parse_ex(&settings, input.c_str(), input.size(), parse_error);
  if (!grammar_json) {
    error_message = string("Invalid JSON at ") + parse_error;
    goto error;
  }

  if (grammar_json->type != json_object) {
    error_message = "Body must be an object";
    goto error;
  }

  name_json = grammar_json->operator[]("name");
  if (name_json.type != json_string) {
    error_message = "Name must be a string";
    goto error;
  }

  name = name_json.u.string.ptr;

  rules_json = grammar_json->operator[]("rules");
  if (rules_json.type != json_object) {
    error_message = "Rules must be an object";
    goto error;
  }

  for (size_t i = 0, length = rules_json.u.object.length; i < length; i++) {
    json_object_entry entry_json = rules_json.u.object.values[i];
    auto result = parse_rule(entry_json.value);
    if (!result.error_message.empty()) {
      error_message = result.error_message;
      goto error;
    }
    grammar.variables.push_back(Variable{
      string(entry_json.name),
      VariableTypeNamed,
      result.rule
    });
  }

  extras_json = grammar_json->operator[]("extras");
  if (extras_json.type != json_none) {
    if (extras_json.type != json_array) {
      error_message = "Extras must be an array";
      goto error;
    }

    for (size_t i = 0, length = extras_json.u.array.length; i < length; i++) {
      json_value *extra_json = extras_json.u.array.values[i];
      auto result = parse_rule(extra_json);
      if (!result.error_message.empty()) {
        error_message = "Invalid extra token: " + result.error_message;
        goto error;
      }
      grammar.extra_tokens.push_back(result.rule);
    }
  }

  conflicts_json = grammar_json->operator[]("conflicts");
  if (conflicts_json.type != json_none) {
    if (conflicts_json.type != json_array) {
      error_message = "Conflicts must be an array";
      goto error;
    }

    for (size_t i = 0, length = conflicts_json.u.array.length; i < length; i++) {
      json_value *conflict_json = conflicts_json.u.array.values[i];
      if (conflict_json->type != json_array) {
        error_message = "Each conflict entry must be an array";
        goto error;
      }

      unordered_set<NamedSymbol> conflict;
      for (size_t j = 0, conflict_length = conflict_json->u.array.length;
           j < conflict_length; j++) {
        json_value *conflict_entry_json = conflict_json->u.array.values[j];
        if (conflict_entry_json->type != json_string) {
          error_message = "Each conflict entry must be an array of strings";
          goto error;
        }

        conflict.insert(rules::NamedSymbol{
          string(conflict_entry_json->u.string.ptr)
        });
      }

      grammar.expected_conflicts.push_back(conflict);
    }
  }

  inline_rules_json = grammar_json->operator[]("inline");
  if (inline_rules_json.type != json_none) {
    if (inline_rules_json.type != json_array) {
      error_message = "Inline rules must be an array";
      goto error;
    }

    for (size_t i = 0, length = inline_rules_json.u.array.length; i < length; i++) {
      json_value *inline_rule_json = inline_rules_json.u.array.values[i];
      if (inline_rule_json->type != json_string) {
        error_message = "Inline rules must be an array of rule names";
        goto error;
      }

      grammar.variables_to_inline.insert(rules::NamedSymbol{string(inline_rule_json->u.string.ptr)});
    }
  }

  external_tokens_json = grammar_json->operator[]("externals");
  if (external_tokens_json.type != json_none) {
    if (external_tokens_json.type != json_array) {
      error_message = "External tokens must be an array";
      goto error;
    }

    for (size_t i = 0, length = external_tokens_json.u.array.length; i < length; i++) {
      json_value *external_token_json = external_tokens_json.u.array.values[i];
      auto result = parse_rule(external_token_json);
      if (!result.error_message.empty()) {
        error_message = "Invalid external token: " + result.error_message;
        goto error;
      }
      grammar.external_tokens.push_back(result.rule);
    }
  }

  word_rule_json = grammar_json->operator[]("word");
  if (word_rule_json.type != json_none) {
    if (word_rule_json.type != json_string) {
      error_message = "Invalid word property";
      goto error;
    }

    grammar.word_token = NamedSymbol { word_rule_json.u.string.ptr };
  }

  json_value_free(grammar_json);
  return { name, grammar, "" };

error:
  if (grammar_json) {
    json_value_free(grammar_json);
  }

  return { "", InputGrammar(), error_message };
}

}  // namespace tree_sitter

File: crates/generate/src/parse_grammar.rs
use rustc_hash::{FxHashMap, FxHashSet};

use regex::Regex;
use serde::{Deserialize, Serialize};
use serde_json::{Map, Value};
use thiserror::Error;

use crate::{
    Diagnostic,
    grammars::{InputGrammar, PrecedenceEntry, ReservedWordContext, Variable},
    rules::{Precedence, Rule, RuleId, RulePool},
    strpool::StrId,
};

#[derive(Deserialize)]
#[serde(tag = "type")]
#[expect(
    non_camel_case_types,
    reason = "variant names match JSON grammar format"
)]
#[expect(
    clippy::upper_case_acronyms,
    reason = "variant names match JSON grammar format"
)]
enum RuleJSON {
    ALIAS {
        content: Box<Self>,
        named: bool,
        value: String,
    },
    BLANK,
    STRING {
        value: String,
    },
    PATTERN {
        value: String,
        flags: Option<String>,
    },
    SYMBOL {
        name: String,
    },
    CHOICE {
        members: Vec<Self>,
    },
    FIELD {
        name: String,
        content: Box<Self>,
    },
    SEQ {
        members: Vec<Self>,
    },
    REPEAT {
        content: Box<Self>,
    },
    REPEAT1 {
        content: Box<Self>,
    },
    PREC_DYNAMIC {
        value: i32,
        content: Box<Self>,
    },
    PREC_LEFT {
        value: PrecedenceValueJSON,
        content: Box<Self>,
    },
    PREC_RIGHT {
        value: PrecedenceValueJSON,
        content: Box<Self>,
    },
    PREC {
        value: PrecedenceValueJSON,
        content: Box<Self>,
    },
    TOKEN {
        content: Box<Self>,
    },
    IMMEDIATE_TOKEN {
        content: Box<Self>,
    },
    RESERVED {
        context_name: String,
        content: Box<Self>,
    },
    EOF,
}

#[derive(Deserialize)]
#[serde(untagged)]
enum PrecedenceValueJSON {
    Integer(i32),
    Name(String),
}

#[derive(Deserialize)]
pub struct GrammarJSON {
    pub name: String,
    rules: Map<String, Value>,
    #[serde(default)]
    precedences: Vec<Vec<RuleJSON>>,
    #[serde(default)]
    conflicts: Vec<Vec<String>>,
    #[serde(default)]
    externals: Vec<RuleJSON>,
    #[serde(default)]
    extras: Vec<RuleJSON>,
    #[serde(default)]
    inline: Vec<String>,
    #[serde(default)]
    supertypes: Vec<String>,
    #[serde(default)]
    word: Option<String>,
    #[serde(default)]
    reserved: Map<String, Value>,
}

pub type ParseGrammarResult<T> = Result<T, ParseGrammarError>;

#[derive(Debug, Error, Serialize, Deserialize)]
pub enum ParseGrammarError {
    #[error("{0}")]
    Serialization(String),
    #[error("Rules in the `extras` array must not contain empty strings")]
    InvalidExtra,
    #[error("Invalid rule in precedences array. Only strings and symbols are allowed")]
    Unexpected,
    #[error("Reserved word sets must be arrays")]
    InvalidReservedWordSet,
    #[error("Grammar Error: Unexpected rule `{0}` in `token()` call")]
    UnexpectedRule(String),
}

impl From<serde_json::Error> for ParseGrammarError {
    fn from(value: serde_json::Error) -> Self {
        Self::Serialization(value.to_string())
    }
}

impl InputGrammar {
    /// Strip unused rules from the grammar and clean up references to them
    /// in the surrounding config. (conflicts, supertypes, inline, extras,
    /// externals, precedences).
    ///
    /// A variable is "used" if it is the start rule, the word token, named in
    /// `extras`/`externals`, or transitively reachable via rule references from
    /// any of the above.
    fn normalize(mut self, diagnostics: &mut Vec<Diagnostic>) -> Self {
        // Compute the used set via forward DFS from the implicit roots
        // (start rule, word_token, refs in extras and externals).
        //
        // Extras count their top-level `NamedSymbol` as a use (so naming
        // a rule directly in `extras` keeps it), but externals do not (the
        // external entry is the rule itself, not a reference to one).
        let used: FxHashSet<StrId> = {
            let by_name: FxHashMap<StrId, RuleId> =
                self.variables.iter().map(|v| (v.name, v.root)).collect();
            let mut visited: FxHashSet<StrId> = FxHashSet::default();
            let mut stack: Vec<StrId> = Vec::new();
            if let Some(first) = self.variables.first() {
                stack.push(first.name);
            }
            if let Some(word) = self.word_name {
                stack.push(word);
            }
            for &root in &self.extra_roots {
                self.pool.collect_referenced_ids(root, false, &mut stack);
            }
            for &root in &self.external_roots {
                self.pool.collect_referenced_ids(root, true, &mut stack);
            }
            for set in &self.reserved_sets {
                for &root in &set.roots {
                    self.pool.collect_referenced_ids(root, false, &mut stack);
                }
            }
            while let Some(name) = stack.pop() {
                if !visited.insert(name) {
                    continue;
                }
                if let Some(&root) = by_name.get(&name) {
                    self.pool.collect_referenced_ids(root, false, &mut stack);
                }
            }
            visited
        };

        for v in &self.variables {
            if !used.contains(&v.name) {
                continue;
            }
            if !self
                .extra_roots
                .iter()
                .any(|&r| self.pool.rule_is_referenced(r, v.name, false))
            {
                continue;
            }
            let inner_root = match self.pool.node(v.root) {
                Rule::Metadata { rule, .. } => rule,
                _ => v.root,
            };
            let matches_empty = match self.pool.node(inner_root) {
                Rule::String(s) => self.pool.resolve(s).is_empty(),
                Rule::Pattern(value, _) => {
                    Regex::new(self.pool.resolve(value)).is_ok_and(|reg| reg.is_match(""))
                }
                _ => false,
            };
            if matches_empty {
                diagnostics.push(Diagnostic::EmptyStringMatch(
                    self.pool.resolve(v.name).to_string(),
                ));
            }
        }

        // Drop unused variables and clean up references to them in the config
        let dropped: Vec<StrId> = self
            .variables
            .iter()
            .filter(|v| !used.contains(&v.name))
            .map(|v| v.name)
            .collect();
        self.variables.retain(|v| used.contains(&v.name));
        for &name in &dropped {
            self.conflict_names.retain(|c| !c.contains(&name));
            self.supertype_names.retain(|&s| s != name);
            self.inline_names.retain(|&s| s != name);
            let pool = &self.pool;
            self.extra_roots
                .retain(|&r| !pool.rule_is_referenced(r, name, true));
            self.external_roots
                .retain(|&r| !pool.rule_is_referenced(r, name, true));
            self.precedence_orderings.retain(|o| {
                !o.iter()
                    .any(|e| matches!(e, PrecedenceEntry::Symbol(s) if *s == name))
            });
            // Prune entries but keep the context: an intentionally-empty
            // reserved-word context is a meaningful marker that rule bodies
            // may reference by name.
            for set in &mut self.reserved_sets {
                set.roots
                    .retain(|&r| !pool.rule_is_referenced(r, name, false));
            }
        }

        self
    }
}

pub(crate) fn parse_grammar(
    input: &str,
    diagnostics: &mut Vec<Diagnostic>,
) -> ParseGrammarResult<InputGrammar> {
    let grammar_json = serde_json::from_str::<GrammarJSON>(input)?;
    let mut pool = RulePool::default();

    let extra_roots =
        grammar_json
            .extras
            .into_iter()
            .try_fold(Vec::<RuleId>::new(), |mut acc, item| {
                let root = pool.parse_rule(item, false, diagnostics)?;
                if let Rule::String(s) = pool.node(root)
                    && pool.resolve(s).is_empty()
                {
                    Err(ParseGrammarError::InvalidExtra)?;
                }
                acc.push(root);
                ParseGrammarResult::Ok(acc)
            })?;

    let external_roots = grammar_json
        .externals
        .into_iter()
        .map(|e| pool.parse_rule(e, false, diagnostics))
        .collect::<ParseGrammarResult<Vec<_>>>()?;

    let mut precedence_orderings = Vec::with_capacity(grammar_json.precedences.len());
    for list in grammar_json.precedences {
        let mut ordering = Vec::with_capacity(list.len());
        for entry in list {
            ordering.push(match entry {
                RuleJSON::STRING { value } => PrecedenceEntry::Name(pool.intern(&value)),
                RuleJSON::SYMBOL { name } => PrecedenceEntry::Symbol(pool.intern(&name)),
                _ => Err(ParseGrammarError::Unexpected)?,
            });
        }
        precedence_orderings.push(ordering);
    }

    let variables = grammar_json
        .rules
        .into_iter()
        .map(|(name, r)| {
            let name = pool.intern(&name);
            Ok(Variable {
                name,
                root: pool.parse_rule(serde_json::from_value(r)?, false, diagnostics)?,
            })
        })
        .collect::<ParseGrammarResult<Vec<_>>>()?;

    let reserved_sets = grammar_json
        .reserved
        .into_iter()
        .map(|(name, rule_values)| {
            let Value::Array(rule_values) = rule_values else {
                Err(ParseGrammarError::InvalidReservedWordSet)?
            };

            let name = pool.intern(&name);
            let mut roots = Vec::with_capacity(rule_values.len());
            for value in rule_values {
                roots.push(pool.parse_rule(serde_json::from_value(value)?, false, diagnostics)?);
            }
            Ok(ReservedWordContext { name, roots })
        })
        .collect::<ParseGrammarResult<Vec<_>>>()?;

    let supertype_names = grammar_json
        .supertypes
        .iter()
        .map(|s| pool.intern(s))
        .collect();
    let conflict_names = grammar_json
        .conflicts
        .iter()
        .map(|c| c.iter().map(|n| pool.intern(n)).collect())
        .collect();
    let inline_names = grammar_json.inline.iter().map(|n| pool.intern(n)).collect();
    let word_name = grammar_json.word.as_deref().map(|w| pool.intern(w));
    let name = pool.intern(&grammar_json.name);

    let grammar = InputGrammar {
        pool,
        name,
        variables,
        external_roots,
        extra_roots,
        reserved_sets,
        supertype_names,
        conflict_names,
        inline_names,
        word_name,
        precedence_orderings,
    }
    .normalize(diagnostics);
    Ok(grammar)
}

impl RulePool {
    fn parse_rule(
        &mut self,
        json: RuleJSON,
        is_token: bool,
        diagnostics: &mut Vec<Diagnostic>,
    ) -> ParseGrammarResult<RuleId> {
        match json {
            RuleJSON::ALIAS {
                content,
                named,
                value,
            } => {
                let content = self.parse_rule(*content, is_token, diagnostics)?;
                let value = self.intern(&value);
                Ok(self.alias(content, value, named))
            }
            RuleJSON::BLANK => Ok(self.blank()),
            RuleJSON::PATTERN { value, flags } => {
                let processed_flags = flags.map_or(String::new(), |f| {
                    f.matches(|c| {
                        if c == 'i' {
                            true
                        } else {
                            // silently ignore unicode flags
                            if c != 'u' && c != 'v' {
                                diagnostics.push(Diagnostic::UnsupportedRegexFlag {
                                    flag: c,
                                    pattern: value.clone(),
                                });
                            }
                            false
                        }
                    })
                    .collect()
                });
                let value = self.intern(&value);
                let flags = self.intern(&processed_flags);
                Ok(self.pattern(value, flags))
            }
            RuleJSON::SYMBOL { name } => {
                if is_token {
                    Err(ParseGrammarError::UnexpectedRule(name))?
                } else {
                    let sid = self.intern(&name);
                    Ok(self.named_symbol(sid))
                }
            }
            RuleJSON::CHOICE { members } => {
                let members = members
                    .into_iter()
                    .map(|m| self.parse_rule(m, is_token, diagnostics))
                    .collect::<ParseGrammarResult<Vec<_>>>()?;
                Ok(self.choice(&members))
            }
            RuleJSON::SEQ { members } => self.try_seq(members.into_iter(), |pool, m| {
                pool.parse_rule(m, is_token, diagnostics)
            }),
            RuleJSON::FIELD { name, content } => {
                let content = self.parse_rule(*content, is_token, diagnostics)?;
                let name = self.intern(&name);
                Ok(self.field(name, content))
            }
            RuleJSON::REPEAT { content } => {
                let content = self.parse_rule(*content, is_token, diagnostics)?;
                let repeat = self.repeat(content);
                let blank = self.blank();
                // Neither element is a `Choice` or equal to one another, so we
                // can skip `self.choice`'s flattening loop.
                let range = self.push_children(&[repeat, blank]);
                Ok(self.push_node(Rule::Choice(range)))
            }
            RuleJSON::REPEAT1 { content } => {
                let content = self.parse_rule(*content, is_token, diagnostics)?;
                Ok(self.repeat(content))
            }
            RuleJSON::PREC { value, content } => {
                let content = self.parse_rule(*content, is_token, diagnostics)?;
                let value = value.intern(self);
                Ok(self.prec(value, content))
            }
            RuleJSON::PREC_LEFT { value, content } => {
                let content = self.parse_rule(*content, is_token, diagnostics)?;
                let value = value.intern(self);
                Ok(self.prec_left(value, content))
            }
            RuleJSON::PREC_RIGHT { value, content } => {
                let content = self.parse_rule(*content, is_token, diagnostics)?;
                let value = value.intern(self);
                Ok(self.prec_right(value, content))
            }
            RuleJSON::PREC_DYNAMIC { value, content } => {
                let content = self.parse_rule(*content, is_token, diagnostics)?;
                Ok(self.prec_dynamic(value, content))
            }
            RuleJSON::RESERVED {
                context_name,
                content,
            } => {
                let content = self.parse_rule(*content, is_token, diagnostics)?;
                let ctx = self.intern(&context_name);
                Ok(self.reserved(content, ctx))
            }
            RuleJSON::TOKEN { content } => {
                let content = self.parse_rule(*content, true, diagnostics)?;
                Ok(self.token(content))
            }
            RuleJSON::IMMEDIATE_TOKEN { content } => {
                let content = self.parse_rule(*content, true, diagnostics)?;
                Ok(self.immediate_token(content))
            }
            RuleJSON::STRING { value } => {
                let sid = self.intern(&value);
                Ok(self.string(sid))
            }
            RuleJSON::EOF => Ok(self.eof()),
        }
    }
}

impl PrecedenceValueJSON {
    fn intern(self, pool: &mut RulePool) -> Precedence {
        match self {
            Self::Integer(i) => Precedence::Integer(i),
            Self::Name(n) => Precedence::Name(pool.intern(&n)),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_parse_grammar() {
        let grammar = parse_grammar(
            r#"{
            "name": "my_lang",
            "rules": {
                "file": {
                    "type": "REPEAT1",
                    "content": {
                        "type": "SYMBOL",
                        "name": "statement"
                    }
                },
                "statement": {
                    "type": "STRING",
                    "value": "foo"
                }
            }
        }"#,
            &mut Vec::new(),
        )
        .unwrap();

        assert_eq!(grammar.pool.resolve(grammar.name), "my_lang");

        let names = grammar
            .variables
            .iter()
            .map(|v| grammar.pool.resolve(v.name))
            .collect::<Vec<_>>();
        assert_eq!(names, ["file", "statement"]);

        // file = repeat(named_symbol "statement")
        let file_child = grammar.pool.node(grammar.variables[0].root);
        let Rule::Repeat(inner) = file_child else {
            panic!("Expected repeat, got {file_child:#?}");
        };
        let inner_node = grammar.pool.node(inner);
        let Rule::NamedSymbol(name) = inner_node else {
            panic!("Expected named symbol, got {inner_node:#?}");
        };
        assert_eq!(grammar.pool.resolve(name), "statement");

        // statement = string "foo"
        let statement_child = grammar.pool.node(grammar.variables[1].root);
        let Rule::String(value) = statement_child else {
            panic!("Expected string, got {statement_child:#?}");
        };
        assert_eq!(grammar.pool.resolve(value), "foo");
    }
}
